### PR TITLE
HTML Renderer

### PR DESCRIPTION
**Pull request by gmac**
This builds out the gem's HTML renderer. This is a much more robust implementation than my initial tinkering in https://github.com/voxmedia/rich-text-ruby/pull/5. The main difference here is that Nokogiri is added so that HTML can be constructed as a real node tree versus as a flat string (which quickly became limiting).

While this isn't a direct rewrite of https://github.com/voxmedia/convert-rich-text, there are definitely similarities.

### CODE REVIEW

File: test/unit/html_test.rb [hunk 1]
```diff
@@ -0,0 +1,331 @@
+require 'test_helper'
+
+describe RichText::HTML do
+
+  it 'renders basic paragraph string' do
+    d = RichText::Delta.new([{ insert: "hello world\n" }])
+    assert_equal '<p>hello world</p>', render_compact_html(d)
```
Comment (saragee3) [inline test/unit/html_test.rb:28]: awesome tests! really cool tool. where in this project are deletions handled?

`{ insert: 'a' }` = `<p>a</p>` but i delete `a` and insert `b` so the it overwrites `<p>a</p>` with `<p>b</p>`?

Comment (gmac) [reply test/unit/html_test.rb:28]: Similar to https://github.com/voxmedia/convert-rich-text, this is an insert-only renderer. You'll see it [intentionally errors](https://github.com/voxmedia/rich-text-ruby/pull/6/files#diff-6c620e9361f3933df560d1ea01e45db7R31) if you try to run it on a transitional delta.

File: lib/rich-text/html.rb [hunk 2]
```diff
@@ -1,66 +1,195 @@
...
 
 module RichText
-  # @todo Work in progress
   class HTML
     ConfigError = Class.new(StandardError)
 
-    attr_reader :config
+    def self.render_xml(delta, options={})
+      new(RichText.config, options).render(delta)
+    end
```
Comment (thomsbg) [inline lib/rich-text/html.rb]: Why separate this from the normal `render`? The internal nokogiri representation seems like an implementation detail I'd rather keep private.

File: lib/rich-text.rb [hunk 3]
```diff
@@ -15,17 +15,38 @@ def self.configure
...
+      el[:class] = classes.uniq.join(' ')
+      el
+    }},
+    secondheader:   { type: :block,  tag: 'h2' },
+    sixthheader:    { type: :block,  tag: 'h6' },
+    small:          { type: :inline, tag: 'small' },
+    strike:         { type: :inline, tag: 's' },
+    subscript:      { type: :inline, tag: 'sub' },
+    superscript:    { type: :inline, tag: 'sup' },
+    thirdheader:    { type: :block,  tag: 'h3' },
```
Comment (thomsbg) [inline lib/rich-text.rb]: This list is growing long and Chorus-specific. I wonder if it would be better to have _no_ formats defined by default and to force consumers to define their own?

File: lib/rich-text.rb [hunk 4]
```diff
@@ -15,17 +15,38 @@ def self.configure
...
+    firstheader:    { type: :block,  tag: 'h1' },
+    fourthheader:   { type: :block,  tag: 'h4' },
+    hr:             { type: :inline, tag: 'hr', block_format: false },
+    id:             { type: :block,  apply: ->(el, op){ el[:id] = op.attributes[:id] } },
+    image:          { type: :inline, tag: 'img', block_format: :figure, apply: ->(el, op){ el[:src] = op.value.dig(:image, :src) } },
+    ins:            { type: :inline, tag: 'ins' },
+    italic:         { type: :inline, tag: 'em' },
+    link:           { type: :inline, tag: 'a', apply: ->(el, op){ el[:href] = op.attributes[:link] } },
+    list:           { type: :block,  tag: 'li', parent: 'ol' },
+    paragraph:      { type: :block,  tag: 'p' },
```
Comment (thomsbg) [inline lib/rich-text.rb]: Including formats in this hash that don't actually correspond to any rich-text attribute key seems a little dangerous and confusing. I like the idea to reuse the same structure for "helper" formats like `div` and `paragraph`, but maybe it would be safer to define those in a separate Hash / config item? Or maybe I'm just being paranoid. 2/10

File: lib/rich-text/html.rb [hunk 5]
```diff
@@ -1,66 +1,195 @@
...
-      new(RichText.config).render(delta)
+    def self.render(delta, options={})
+      render_xml(delta, options).inner_html
     end
 
-    def initialize(config)
-      @default_block_tag = config.html_default_block_tag
-      @block_tags = config.html_block_tags
-      @inline_tags = config.html_inline_tags
+    attr_reader :doc, :root
```
Comment (thomsbg) [inline lib/rich-text/html.rb]: Again, `doc` and `root` seem like internal implementation details of this class, and something I'd rather keep private. Using plain instance variable references for `@doc` and `@root` seems better.

File: lib/rich-text/html.rb [hunk 6]
```diff
@@ -1,66 +1,195 @@
...
+
+    def initialize(config, options)
+      @default_block_format = options[:default_block_format] || config.html_default_block_format
+      @formats = config.html_formats.merge(options[:formats] || {})
+      @inline_formats = Hash[@formats.select { |k, v| !v.key?(:type) || v[:type] == :inline }]
+      @block_formats = Hash[@formats.select { |k, v| v[:type] == :block }]
```
Comment (thomsbg) [inline lib/rich-text/html.rb]: If these end up being split apart anyway, why not define them in config separately? `config.html_block_formats` & `config.html_inline_formats`.

File: lib/rich-text/html.rb [hunk 7]
```diff
@@ -1,66 +1,195 @@
...
+      attributes ||= op.attributes
+
+      # order of operations for rendering formats
+      # tag formats are applied first (sorted by priority)
+      # other attribute formats follow (sorted by priority)
+      ordered_formats = attributes.keys.map {|k| formats[k.to_sym] }.compact.sort do |a, b|
+        a = [a.key?(:tag) ? 0 : 1, a[:priority] || Float::INFINITY]
+        b = [b.key?(:tag) ? 0 : 1, b[:priority] || Float::INFINITY]
+        a <=> b
+      end
```
Comment (thomsbg) [inline lib/rich-text/html.rb:142]: How did you settle on the default ordering rule for applying formats with tags first? Would things work out OK if we simply applied formats in alphabetical order by key name, instead of this priority system? Doesn't look like the default config uses any `:priority` values...

File: test/unit/html_test.rb [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+    d = RichText::Delta.new([
+      { insert: 'mali principii' },
+      { insert: "\n" },
+      { insert: 'malus finis', attributes: { invalid: true } }
+    ])
+    assert_equal '<p>mali principii</p><p>malus finis</p>', render_compact_html(d)
+  end
+
+  def render_compact_html(delta, options={})
+    RichText::HTML.render_xml(delta, options).inner_html(save_with: 0)
```
Comment (thomsbg) [inline test/unit/html_test.rb]: What does `save_with: 0` do here? Change the way whitespace is handled? I can't find good docs on that option from Nokogiri.

**Review by thomsbg — COMMENTED** [review]
Great stuff! tbh, I found the new code in the `HTML` class a bit hard to follow, and I was tempted to see if I could refactor it in order to make sense of it, but I ran out of time.

One idea I had: would it make things simpler if inline formatting and block formatting were applied in two separate passes through the document? i.e. One pass to apply inline formatting inside of default block tags, and a second pass to modify block tags as needed. I was thinking that might make it easier to write the logic for things like merging adjacent `list` items, or joining adjacent lines with for the `br` format.

File: lib/rich-text.rb (hunk 3 above)
Comment (saragee3) [reply lib/rich-text.rb]: or maybe give the option for them to define their own?

File: lib/rich-text/html.rb (hunk 2 above)
Comment (gmac) [reply lib/rich-text/html.rb]: I see a lot of value in getting the node tree from the library, versus getting the HTML string and having to immediately reparse it when what you wanted was a document structure.

File: lib/rich-text.rb (hunk 3 above)
Comment (gmac) [reply lib/rich-text.rb]: I'm fine with paring it down; the formats themselves (and supporting tests) could move directly over into SBN. That said, seems like there should be _some_ simple formats so that the thing does something out of the box. Overriding the entire default collection with your own is really simple.

File: lib/rich-text.rb (hunk 4 above)
Comment (gmac) [reply lib/rich-text.rb]: I guess I took the approach of: this is a definition of stuff that can show up in your rendered output. If it intersects with your schema, use a different name for the helper format. Either way though, I'd want the renderer to produce artifacts explicitly as I'd spec'd them.

File: lib/rich-text/html.rb (hunk 5 above)
Comment (gmac) [reply lib/rich-text/html.rb]: Like the above point – what if my goal in using this is to produce an XML document for other purposes? Or I want to change the encoding, or do any of the other stuff you can do with an XML document? Making those public supports advanced uses with no complexity to added to the basic workflow – which would predominantly be `delta.to_html`. IMHO

File: lib/rich-text/html.rb (hunk 6 above)
Comment (gmac) [reply lib/rich-text/html.rb]: Fair question – I was going back and forth on that and just happened to end in this state. The major case for this direction would be having an `all` type for stuff like attribute setters that could apply to either blocks or inline elements. Partitioning the sets would require two definitions – maybe that's not the worst thing?

File: lib/rich-text/html.rb (hunk 7 above)
Comment (gmac) [reply lib/rich-text/html.rb:142]: Order is important when evaluating tag versus attribute formats (or – "other" formats that just apply an element modifier). Tag formats need to be rendered first so that there's something there for the attribute/other format to target. Priority is just an added courtesy that would allow tag order to be weighted if you cared about such things – I don't have a specific case where you would, but it seems plausible.

File: test/unit/html_test.rb (hunk 8 above)
Comment (gmac) [reply test/unit/html_test.rb]: Yep, that mysterious param is apparently the Nokogiri squish option for rendering without any pretty formatting... I found it on a stack overflow post and it did what I wanted. It's a terrible option name, right?

**Comment by gmac** [comment]
@thomsbg re:

> Great stuff! tbh, I found the new code in the HTML class a bit hard to follow, and I was tempted to see if I could refactor it in order to make sense of it, but I ran out of time.

This implementation is a bit different from (my best understanding of) the https://github.com/voxmedia/convert-rich-text in two ways:

1. It's designed to render the final document structure in a single pass. This approach was somewhat by design, and somewhat by necessity given that Nokogiri doesn't do all the DOM callbacks that allow rules to be written around future parents. I also wanted to avoid the post-processing route because tags from many different ops could all look the same at the end. If you have a global post-process you want to run, you could always grab the builder's Nokogiri doc and go to town on it.

2. I also deliberately wrote some aspects to be a little fuzzier for the sake of fault tolerance. While some rules _should_ be clean cut, I think it gives better results when the renderer makes some corrections for missing newlines.

File: lib/rich-text/html.rb (hunk 5 above)
Comment (gmac) [reply lib/rich-text/html.rb]: I've removed root but left doc. It has value if you're trying to use this as an isolated tool.

File: lib/rich-text.rb (hunk 4 above)
Comment (gmac) [reply lib/rich-text.rb]: I've addresses this so that block formats intended as structural helpers don't get definitions. So, elements found in the delta will not respond to those helper formats.

File: lib/rich-text.rb (hunk 3 above)
Comment (gmac) [reply lib/rich-text.rb]: I've moved the bulk of the definitions and their corresponding tests over the SBN in this branch here: https://github.com/voxmedia/sbn/compare/gm-richtext-formats?expand=1

What's left in this gem are a skeleton set of basic rules so that it does _something_ out of the box. Also, the tests for these formats have been pared down – tests now focus on specific  implementation details and less on general formats. I still need to write README documentation about configuring HTML rules, and will have that in by EOD tomorrow.

File: lib/rich-text/html.rb (hunk 2 above)
Comment (gmac) [reply lib/rich-text/html.rb]: Down to just `render`. Power users can go through `.new(...)`.

**Comment by gmac** [comment]
Ready for another look @thomsbg – this has been pared down to skeletal gem mechanics.

**Comment by gmac** [comment]
Supporting documentation can be read here: https://github.com/voxmedia/rich-text-ruby/tree/gm-structured-html#html-formatting

File: rich-text.gemspec [hunk 9]
```diff
@@ -22,7 +22,8 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency "diff-lcs", "~> 1.2.5"
   spec.add_dependency "activesupport", ">= 3.0.0"
-  # spec.add_dependency "nokogiri", ">= 1.0.0"
+  spec.add_dependency "nokogiri", ">= 1.10.8"
```
Comment (thomsbg) [inline rich-text.gemspec]: It will make people's lives easier if we were as relaxed about this dependency as possible. Are you using any specific feature of nokogiri that's not available in previous 1.x versions?

**Review by thomsbg — APPROVED** [review]
